h5netcdf/core.py

```diff
@@ -125,28 +132,61 @@ def shuffle(self):
 
 NOT_A_VARIABLE = b'This is a netCDF dimension but not a netCDF variable.'
 
+class lazy_objects(Mapping):
+    def __init__(self, object_cls):
+        self._objects = list()
+        self._loaded_objects = dict()
+        self._object_cls = object_cls
+        return
+
+    def set(self, name, object):
+        self._objects.append(name)
+        self._loaded_objects[name] = object
+        return
+
+    def add(self, name):
+        self._objects.append(name)
+        return
+
+    def __iter__(self):
+        for name in self._objects:
+            yield name
+
+    def __len__(self):
+        return len(self._objects)
+
+    def __getitem__(self, key):
+        if not key in self._objects:
+            raise KeyError('Object {0} is not avaialble.'.format(key))
+        elif key in self._loaded_objects.keys():
+            return self._loaded_objects[key]
+        else:
+            self._loaded_objects[key] = self._object_cls(key)
+            return self[key]
 
 class Group(Mapping):
+
     _variable_cls = Variable
 
     @property
     def _group_cls(self):
         return Group
 
-    def __init__(self, parent, h5group):
+    def __init__(self, parent, name):
         self._parent = parent
         self._root = parent._root
-        self._h5group = h5group
+        self._h5path = '/'.join([parent.name.rstrip('/'), name.lstrip('/')])
+
         if parent is not self:
             self._dim_sizes = parent._dim_sizes.new_child()
             self._dim_order = parent._dim_order.new_child()
 
-        self._variables = OrderedDict()
-        self._groups = OrderedDict()
+        self._variables = lazy_objects(lambda x: self._variable_cls(self, x))
```

Comment:
I would change the lazy_objects constructor to take two arguments. It's just a little cleaner than using the lambda.